### PR TITLE
feat: sync unidirectional relations

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Updates uni directional relations to target the right entries when overriding published or draft entries.

Based on the following [notion RFC](https://www.notion.so/strapi/Unidirectional-relations-on-publish-discard-29d0133709c646b088f377727c9c7ff1)

### CODE REVIEW

File: packages/core/core/src/services/document-service/repository.ts [hunk 1]
```diff
@@ -304,18 +308,20 @@ export const createContentTypeRepository: RepositoryFactoryMethod = (uid) => {
...
+        select: ['id', 'locale'],
       }),
     ]);
 
-    // Delete all drafts
-    await async.map(versionsToDelete, (entry: any) => entries.delete(entry.id));
-
     // Transform published entry data and create draft versions
-    const draftEntries = await async.map(versionsToDraft, (entry: any) =>
+    const draftEntries = await async.map(versionsToDraft, async (entry: any) =>
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/repository.ts:316]: Just wondering if it's actually needed ?

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  oldEntries: { id: string; locale: string }[],
+  newEntries: { id: string; locale: string }[]
+) => {
+  const relationsToSync: RelationToSync[] = [];
+
+  // Iterate all components and content types
+  const contentTypes = Object.values(strapi.contentTypes) as Schema.ContentType[];
+  const components = Object.values(strapi.components) as Schema.Component[];
+
+  for (const contentType of [...contentTypes, ...components]) {
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:30]: Maybe use another name than contentType

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    // If the content type has a relation to the current content type
+    for (const [name, attribute] of Object.entries(contentType.attributes) as any) {
+      if (attribute.type !== 'relation') continue;
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:33]: Would it be safer to target the specific oneWay & manyWay relational types directly ?

File: packages/core/core/src/services/document-service/repository.ts [hunk 4]
```diff
@@ -250,15 +251,18 @@ export const createContentTypeRepository: RepositoryFactoryMethod = (uid) => {
...
 
+    await syncUnidirectionalRelations(uid, oldPublishedVersions, publishedEntries);
+
+    // Delete old published versions
+    await async.map(oldPublishedVersions, (entry: any) => entries.delete(entry.id));
+
     publishedEntries.forEach(emitEvent('entry.publish'));
+
     return { documentId, entries: publishedEntries };
   }
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/repository.ts:268]: More of a general thought remark, when deleting data we do not clear the unidirectional relations pointing to nothing anymore. Sth we need to think about as a "cleanup task done in the bg"

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      if (!oldEntry) return;
+
+      // Update all relations to the new entry
+      for (const { uid, attribute: name } of relationsToSync) {
+        const model = strapi.db.metadata.get(uid);
+        const attribute = model.attributes[name];
+        if (attribute?.type === 'relation') {
+          // @ts-expect-error - FIX
+          const joinTable = attribute.joinTable;
+          if (!joinTable) continue;
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:55]: if there is a joinColumn we should also handle it I think

We def should look into adding some higher level db queries to make this simpler 🤔

**Review by alexandrebodin — COMMENTED** [review]
The logic looks sane and well encapsulated. Still those low level queries I think we have to offer an api for instead but not an issue for this PR

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts (hunk 3 above)
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:33]: I don't fully understand the question, isn't this bit only selecting the one way/many way relations ? (by checking if the attr has inversedBy & mappedBy)

File: packages/core/core/src/services/document-service/repository.ts (hunk 4 above)
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/repository.ts:268]: can we rely on the cascade delete we currently have?

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts (hunk 5 above)
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:55]: good point, will implement this for joinColumns too! 

do we currently have any db api to more easily update joinTables atm? ideally I would prefer to have a batch update on relations but that would be too much to implement atm

File: packages/core/core/src/services/document-service/repository.ts (hunk 4 above)
Comment (alexandrebodin) [reply packages/core/core/src/services/document-service/repository.ts:268]: in some cases yes but not for all db or polymorphics)

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts (hunk 5 above)
Comment (alexandrebodin) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:55]: we don't but you got the idea right for later ;)

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts (hunk 3 above)
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:33]: do you know of other ways of filtering by one way relations? I was checking the available metadata on the attribute and we only store the following:
```
{
  type: "relation",
  relation: "oneToOne",
  target: "api::tag.tag",
}
```

Comment (alexandrebodin) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:33]: Ok that's my bad, we call them oneWay & manyWay only inside the ctb 🤦

Comment (derrickmehaffy) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:33]: Trust me I do that too 😂

**Comment by Marc-Roig** [comment]
@alexandrebodin we still do not support publishing entries with join columns, it's something I can work next, what do you think if we merge this one first and I work on that on a follow PR?

**Review by alexandrebodin — APPROVED** [review]
Code LGTM. I didn't QA, let me know if you need help with that

**Comment by Marc-Roig** [comment]
@alexandrebodin a bit less compact, but I have updated the code to load+update unidirectional relations instead of updating them in a single query.

The original approach made the unique validations fail, we were not deleting the old published versions before creating the new ones. Hence triggering those validation errors

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  oldEntries: { id: string; locale: string }[],
+  newEntries: { id: string; locale: string }[],
+  oldRelations: { joinTable: any; relations: any[] }[]
+) => {
+  /**
+   * Create a map of old entry ids to new entry ids
+   *
+   * Will be used to update the relation target ids
+   */
+  const newEntryByLocale = keyBy('locale', newEntries);
```
Comment (alexandrebodin) [inline packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:77]: We shouldn't need to handle the case with no locale or locale being null but maybe worth some check 🤔

Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:77]: will make it a bit more robust to that & will set up a test, last time I checked the locale value was always null on that scenario

**Review by alexandrebodin — APPROVED** [review]
This looks good 👍
